src/middleware/routeHandler.ts

```diff
@@ -0,0 +1,250 @@
+    next(err)
+  })
+
+  try {
+    await unlock(siteName)
+  } catch (err) {
+    next(err)
+  }
+}
+
+export const attachRollbackRouteHandlerWrapper = (routeHandler: any) => async (
+  req: any,
+  res: any,
+  next: any
+) => {
+  const { userSessionData } = res.locals
+  const { siteName } = req.params
+
+  const { accessToken } = userSessionData
+  const { growthbook } = req
+
+  const shouldUseGitFileSystem = !!growthbook?.getFeatureValue(
+    FEATURE_FLAGS.IS_GGS_ENABLED,
+    false
+  )
+
+  const isGitAvailable = await handleGitFileLock(siteName, next)
+  if (!isGitAvailable) return
+  try {
+    await lock(siteName)
+  } catch (err) {
+    next(err)
+    return
+  }
+
+  let originalStagingCommitSha: any
+  let originalStagingLiteCommitSha: any
+
+  if (shouldUseGitFileSystem) {
+    const results = await Promise.all([
+      gitFileSystemService.getLatestCommitOfBranch(siteName, STAGING_BRANCH),
+      gitFileSystemService.getLatestCommitOfBranch(
+        siteName,
+        STAGING_LITE_BRANCH
+      ),
+    ])
+    const [stagingResult, stagingLiteResult] = results
+
+    if (stagingResult.isErr() || stagingLiteResult.isErr()) {
+      await unlock(siteName)
+      if (stagingResult.isErr()) next(stagingResult.error)
+      if (stagingLiteResult.isErr()) next(stagingLiteResult.error)
+      return
+    }
+    originalStagingCommitSha = stagingResult.value.sha
+    originalStagingLiteCommitSha = stagingLiteResult.value.sha
+    if (!originalStagingCommitSha || !originalStagingLiteCommitSha) {
+      await unlock(siteName)
+      if (stagingResult.isErr()) next(stagingResult.error)
+      return
+    }
+    // Unused for git file system, but to maintain existing structure
+    res.locals.githubSessionData = new GithubSessionData({
+      currentCommitSha: "",
+      treeSha: "",
+    })
+  } else {
+    try {
+      const {
+        currentCommitSha: currentStgCommitSha,
+        treeSha: stgTreeSha,
+      } = await getCommitAndTreeSha(siteName, accessToken, STAGING_BRANCH)
+
+      const {
+        currentCommitSha: currentStgLiteCommitSha,
+      } = await getCommitAndTreeSha(siteName, accessToken, STAGING_LITE_BRANCH)
+
+      const githubSessionData = new GithubSessionData({
+        currentCommitSha: currentStgCommitSha,
+        treeSha: stgTreeSha,
+      })
+      res.locals.githubSessionData = githubSessionData
+
+      originalStagingCommitSha = currentStgCommitSha
+      originalStagingLiteCommitSha = currentStgLiteCommitSha
+    } catch (err) {
+      await unlock(siteName)
+      next(err)
+      return
+    }
+  }
+
+  await routeHandler(req, res, next).catch(async (err: any) => {
+    try {
+      if (shouldUseGitFileSystem) {
+        await backOff(() =>
+          convertNeverThrowToPromise(
+            gitFileSystemService.rollback(
+              siteName,
+              originalStagingCommitSha,
+              STAGING_BRANCH
+            )
+          )
+        )
+
+        await backOff(() =>
```

Comment:
how many times does this retry?

Reply:
max default 10

Reply:
1st attempt
200ms
2nd
400ms
3rd
800ms
4th
1600ms 
5th
3200ms
7th
6400ms

and so on right? is the above backoff sequence correct?

was just thinking in a worst case scenario the request will be open for quite long? will this timeout on client?

Reply:
would need to dive into source code here to get the exact algo (https://github.com/coveooss/exponential-backoff), but based on the doc i would assume that would be the case

wait ah this backoff occurs when the call has an error bah, so the FE would have already gotten an error code that we try to fix via exponential backoff

Reply:
hmm this is a handler middleware right? so it will completed through the execution of all handlers before request completes right?

Reply:
for clarity ya, this is in the .catch() control flow, would this not mean that this occurs after a certain error short circuits the entire chain of executions?

Reply:
follow up suggestion, option 1 + alarm if it fails for 5 times as failing 5 times signifies some deeper issue not resolvable with retry

Reply:
sure option 1 [implemented](https://github.com/isomerpages/isomercms-backend/pull/1044/commits/f626ef7dd6ad56130ff3dfe6191878f2538bf72a)

Reply:
hmm sure about alarm? worried about numberous false positives tho
What I can do is add logging for this, can review later if this occurs too freq with false positives. will add todo regardless in jira

Reply:
okay

Reply:
@kishore03109 pls test once on staging/local